lib/logstash/outputs/newrelic.rb

```diff
@@ -116,17 +118,52 @@ def multi_receive(events)
 
   def handle_response(response)
     if !(200 <= response.code.to_i && response.code.to_i < 300)
-      @logger.error("Request returned " + response.code + " " + response.body)
+      raise Error::BadResponseCodeError.new(
+        response.code.to_i, @base_uri, response.message
+      )
     end
   end
 
   def nr_send(payload)
-    http = Net::HTTP.new(@end_point.host, 443)
-    request = Net::HTTP::Post.new(@end_point.request_uri)
-    http.use_ssl = true
-    http.verify_mode = OpenSSL::SSL::VERIFY_PEER
-    @header.each {|k, v| request[k] = v}
-    request.body = payload
-    handle_response(http.request(request))
+    retries = 0
+    begin
+      http = Net::HTTP.new(@end_point.host, 443)
+      request = Net::HTTP::Post.new(@end_point.request_uri)
+      http.use_ssl = true
+      http.verify_mode = OpenSSL::SSL::VERIFY_PEER
+      @header.each { |k, v| request[k] = v }
+      request.body = payload
+      handle_response(http.request(request))
+    rescue Error::BadResponseCodeError => e
+      @logger.error(e.message)
+      if (should_retry(retries) && is_retryable_code(e))
+        retries += 1
+        sleep(1)
+        retry
+      end
+    rescue => e
+      # Stuff that should never happen
+      # For all other errors print out full issues
+      @logger.error(
+        "An unknown error occurred sending a bulk request to NewRelic.",
+        :error_message => e.message,
+        :error_class => e.class.name,
+        :backtrace => e.backtrace
+      )
+      if (should_retry(retries))
+        retries += 1
+        sleep(1)
+        retry
+      end
+    end
+  end
+
+  def should_retry(retries)
+    @enable_retry && retries < @max_retry
+  end
+
+  def is_retryable_code(response_error)
+    error_code = response_error.response_code
+    error_code != 401 && error_code != 403
```

Comment:
I would suggest defining a constant list, at the top of the class, named `RETRIABLE_CODES` or similar, and then I would check  here whether the returned code is contained in this list.